pkg/controller/build/buildrequest/machineosbuild.go

```diff
@@ -0,0 +1,244 @@
+	ctrlcommon "github.com/openshift/machine-config-operator/pkg/controller/common"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+)
+
+// This is the same salt / pattern from pkg/controller/render/hash.go
+var (
+	// salt is 80 random bytes.
+	// The salt was generated by `od -vAn -N80 -tu1 < /dev/urandom`. Do not change it.
+	salt = []byte{
+		16, 124, 206, 228, 139, 56, 175, 175, 79, 229, 134, 118, 157, 154, 211, 110,
+		25, 93, 47, 253, 172, 106, 37, 7, 174, 13, 160, 185, 110, 17, 87, 52,
+		219, 131, 12, 206, 218, 141, 116, 135, 188, 181, 192, 151, 233, 62, 126, 165,
+		64, 83, 179, 119, 15, 168, 208, 197, 146, 107, 58, 227, 133, 188, 238, 26,
+		33, 26, 235, 202, 32, 173, 31, 234, 41, 144, 148, 79, 6, 206, 23, 22,
+	}
+)
+
+// Holds the objects that are used to construct a MachineOSBuild with a hashed
+// name.
+type MachineOSBuildOpts struct {
+	MachineOSConfig   *mcfgv1alpha1.MachineOSConfig
+	MachineConfigPool *mcfgv1.MachineConfigPool
+	OSImageURLConfig  *ctrlcommon.OSImageURLConfig
+}
+
+// Validates that the required options are provided.
+func (m *MachineOSBuildOpts) validateForHash() error {
+	if m.MachineOSConfig == nil {
+		return fmt.Errorf("missing required MachineOSConfig")
+	}
+
+	if m.MachineConfigPool == nil {
+		return fmt.Errorf("missing required MachineConfigPool")
+	}
+
+	if m.MachineConfigPool.Name != m.MachineOSConfig.Spec.MachineConfigPool.Name {
+		return fmt.Errorf("name mismatch, MachineConfigPool has %q, MachineOSConfig has %q", m.MachineConfigPool.Name, m.MachineOSConfig.Spec.MachineConfigPool.Name)
+	}
+
+	if m.OSImageURLConfig == nil {
+		return fmt.Errorf("misssing OSImageURLConfig")
+	}
+
+	return nil
+}
+
+// Makes a deep-copy of the MachineOSConfig and clears the data from it that
+// can come from either the MachineOSConfig or from the OSImageURLConfig
+// object. We do this to stabilize the hashing of the name so that whether the
+// value comes from the MachineOSConfig or the OSImageURLConfig, the hash will
+// be same, provided that the value is the same.
+func (m *MachineOSBuildOpts) getMachineOSConfigForHashing() *mcfgv1alpha1.MachineOSConfig {
+	moscCopy := m.MachineOSConfig.DeepCopy()
+	moscCopy.Spec.BuildInputs.BaseOSImagePullspec = ""
+	moscCopy.Spec.BuildInputs.BaseOSExtensionsImagePullspec = ""
+	moscCopy.Spec.BuildInputs.ReleaseVersion = ""
+	return moscCopy
+}
+
+// Creates a list of objects that are consumed by the SHA256 hash.
+func (m *MachineOSBuildOpts) objectsForHash() []interface{} {
+	o := BuildRequestOpts{
+		MachineOSConfig:  m.MachineOSConfig,
+		OSImageURLConfig: m.OSImageURLConfig,
+	}
+
+	// The objects considered for hashing described inline:
+	out := []interface{}{
+		// The spec of the MachineConfigPool object. This includes the name of the
+		// rendered MachineConfig as well as the reference of all of the individual
+		// MachineConfigs that went into that rendered MachineConfig.
+		m.MachineConfigPool.Spec,
```

Comment:
non-blocking clarifying question:

What extra fields will we benefit from in the MCP other than the [mcp.Configuration](https://github.com/openshift/api/blob/4f6053f954b069f7fea5788d4307f4213db53678/machineconfiguration/v1/types.go#L578)? Would it be fair to re-trigger a build if the other fields of the MCP change?

Reply:
I opted only to consider the MachineConfigPool spec because that's what will change the most. It covers the case where a new rendered MachineConfig is created. I don't think it makes sense to cover the status or objectmeta portions of the MachineConfigPool struct since status changes, annotation / label changes, etc. probably should not be considered. I thought about also considering the rendered MachineConfig itself but ultimately, I don't think it's necessary since its name should change based upon its contents.

Reply:
I was proposing if instead of the entire `m.MachineConfigPool.Spec`, we just keep `m.MachineConfigPool.Spec.Configuration`